File: lib/json_api/book.rb
module JsonApi::Book
  extend JsonApi::Json
  
  TYPE_KEY = 'book'
  DEFAULT_PAGE = 25
  MAX_PAGE =50
    
  def self.build_json(book, args={})
    json = {}
    json['id'] = book.ref_id
    json['locale'] = book.locale || 'en'
    json['pending'] = !book.id
    json['approved'] = !!book.approved
    json['public'] = !!book.public
    if book.data
      json['title'] = book.data['title']
      json['author'] = book.data['author']
      json['related_words'] = book.data['related_words'] || book.related_page_words
      json['total_pages'] = book.data['pages'].length
    end
    json['image'] = book.data['image']
    
    if args.has_key?(:permissions) && book.data
      json['permissions'] = book.permissions_for(args[:permissions])
      # other data to include
      Book::STRING_PARAMS.each do |param|
        json[param] = book.data[param].to_s if book.data[param]
      end
      Book::OBJ_PARAMS.each do |param|
        json[param] = book.data[param] if book.data[param]
      end
      json['pages'].each do |page|
        page['target_words'] = page['related_words'].split(/,|\n/).map(&:strip) if page['related_words']
      end
    end
    json
  end
end

File: app/frontend/app/models/book.js
import Ember from 'ember';
import DS from 'ember-data';

var Book = DS.Model.extend({
  didLoad: function() {
    var _this = this;
    if(!this.get('pages') && this.get('pending')) {
      this.set('pages', []);
    }
  },
  pending: DS.attr('boolean'),
  permissions: DS.attr('raw'),
  pages: DS.attr('raw'),
  image: DS.attr('raw'),
  locale: DS.attr('string'),
  related_words: DS.attr('raw'),
  total_pages: DS.attr('number'),
  title: DS.attr('string'),
  book_type: function() {
    return "communication_workshop";
  }.property(),
  author: DS.attr('string'),
  new_core_words: DS.attr('string'),
});

export default Book;
